walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -49,93 +58,120 @@
+    sortedEvents.sort(Event.EVENT_COMPARATOR);
+
+    LinkedList<Event> applicableEventsOnly = removeEventsWithNoAttendees(mandatoryAttendees, sortedEvents);
+    LinkedList<Event> mandatoryNonOverlapping = sortedEventsIntoNonOverlappingSortedEvents(mandatoryAttendees, applicableEventsOnly);
+    LinkedList<TimeRange> available = availableTimesFromSortedNonOverlappingEvents(mandatoryNonOverlapping, request);
 
     if (optionalAttendees.size() == 0) {
       return available;
     }
 
-    ArrayList<TimeRange> availableWithOptional = availableTimesFromSortedEvents(optionalAttendees, allEvents, request);
+    LinkedList<Event> applicableEventsOnlyOfOptional = removeEventsWithNoAttendees(optionalAttendees, sortedEvents);
+    LinkedList<Event> optionalNonOverlapping = sortedEventsIntoNonOverlappingSortedEvents(optionalAttendees, applicableEventsOnlyOfOptional);
+    LinkedList<TimeRange> availableWithOptional = availableTimesFromSortedNonOverlappingEvents(optionalNonOverlapping, request);
     
     if (available.size() == 0) {
       return availableWithOptional;
     }
     
-    ArrayList<TimeRange> finalTimeRanges = new ArrayList<>();
+    LinkedList<TimeRange> finalTimeRanges = mergeTimeRangesThatIntersect(availableWithOptional, available);
+    return finalTimeRanges;
+  }
+   
+  /*
+   * Function to merge time ranges that intersect between two lists of time ranges
+   * 
+   * Runtime: O(n*logm) where 'n' is the length of optional and 'm' is the length of mandatory
+   * 
+   * @param optional    A list of time ranges in which for every one we'll search for an intersection
+   * @param mandatory   A list of time ranges to search within
+   * 
+   * return   A list of time ranges where all ranges are within both input lists
+   * 
+  */
+  private LinkedList<TimeRange> mergeTimeRangesThatIntersect(LinkedList<TimeRange> optional, LinkedList<TimeRange> mandatory) {
+
+    LinkedList<TimeRange> finalTimeRanges = new LinkedList<>();
 
-    for (TimeRange optionalTimeRange: availableWithOptional) {
+    LinkedList<Integer> startValues = new LinkedList<>();
+    for (TimeRange mandatoryTimeRange: mandatory) {
+      startValues.add(mandatoryTimeRange.start());
+    }
+
+    for (TimeRange optionalTimeRange: optional) {
       int start = optionalTimeRange.start();
       int end = optionalTimeRange.end();
 
-      int indexOfIntersection = binarySearchOverlappingTimeRange(available, 0, available.size() - 1, start);
-
-      if (indexOfIntersection != -1) {
-        int mandatoryEnd = available.get(indexOfIntersection).end();
+      // This only works because java's binary search returns (-(insertion_point) - 1) 
+      // if not found, meaning that the negative is 2 indexes beyond what the closest 
+      // lesser than element exists
+      int index = Collections.binarySearch(startValues, start);
+      index = index >= 0 ? index : -index - 2;
```

Comment:
ok, I didn't think too hard about this, but based of (-(insertion_point) -1), shouldn't this be the reverse of those operations?

`insertion_point = (index + 1) * -1`

Reply:
I rewrote it like that but my testing points that it is + 2